### PR TITLE
Add plan adapters and plan persistence manager

### LINKED ISSUE #37: Create PlanPersistenceController

### PR DESCRIPTION
**Pull request by keanecjy**

### CODE REVIEW

File: TourMate/TourMate/Controllers/PlanPersistenceController.swift [hunk 1]
```diff
@@ -5,6 +5,43 @@
...
+struct PlanPersistenceController: PlanPersistenceControllerProtocol {
+
+    let firebasePersistenceManager = FirebasePersistenceManager(collectionId: FirebaseConfig.planCollectionId)
+
+    func addPlan(plan: Plan) async -> (Bool, String) {
+        await firebasePersistenceManager.addItem(id: plan.id, item: PlanAdapter.toAdaptedPlan(plan: plan))
+    }
+
+    func fetchPlan(withPlanId planId: String) async -> (Plan, String) {
+        let (adaptedPlan, errorMessage) = await firebasePersistenceManager.fetchItem(id: planId)
```
Comment (Vielheim) [inline TourMate/TourMate/Controllers/PlanPersistenceController.swift]: Will we need to check if the adaptedPlan Is nil? Or are we letting preconditionFailure() handle it?

File: TourMate/TourMate/Models/DomainModels/Plan.swift [hunk 2]
```diff
@@ -10,10 +10,10 @@ import Foundation
 protocol Plan {
     var id: String { get }
     var tripId: String { get }
-    var planType: PlanType { get set }
+    var planType: PlanType { get }
     var name: String { get set }
     var startDate: Date { get set }
-    var endDate: Date? { get set }
+    var endDate: Date { get set }
```
Comment (tanruiquan) [inline TourMate/TourMate/Models/DomainModels/Plan.swift:16]: Any reason why this was changed? We made it optionally as `Restaurant` does not have an `endDate`

File: TourMate/TourMate/Controllers/PlanPersistenceController.swift [hunk 3]
```diff
@@ -5,6 +5,43 @@
...
+        guard let adaptedPlan = adaptedPlan as? FirebaseAdaptedPlan else {
+            preconditionFailure()
+        }
+
+        let plan = PlanAdapter.toPlan(adaptedPlan: adaptedPlan)
+        return (plan, errorMessage)
+    }
```
Comment (tanruiquan) [inline TourMate/TourMate/Controllers/PlanPersistenceController.swift]: Maybe something like this?

```suggestion
    func fetchPlan(withPlanId planId: String) async -> (Plan?, String) {
        let (adaptedPlan, errorMessage) = await firebasePersistenceManager.fetchItem(id: planId)
        
        guard let adaptedPlan != nil,
          let adaptedPlan = adaptedPlan as? FirebaseAdaptedPlan,
          let errorMessage.isEmpty else {
                     return (nil, errorMessage)
                   }
        let plan = PlanAdapter.toPlan(adaptedPlan: adaptedPlan)
        return (plan, errorMessage)
    }
```

**Review by Vielheim — APPROVED** [review]
I think apart from error handling in PlanPersistenceController, LGTM!

**Review by tanruiquan — APPROVED** [review]
I assume we are sticking to classes for our model? LGTM 👍

File: TourMate/TourMate/Controllers/PlanPersistenceController.swift (hunk 3 above)
Comment (Vielheim) [reply TourMate/TourMate/Controllers/PlanPersistenceController.swift]: Yup, was thinking something like this

File: TourMate/TourMate/Controllers/PlanPersistenceController.swift (hunk 1 above)
Comment (keanecjy) [reply TourMate/TourMate/Controllers/PlanPersistenceController.swift]: Good catch, thanks!

File: TourMate/TourMate/Models/DomainModels/Plan.swift (hunk 2 above)
Comment (keanecjy) [reply TourMate/TourMate/Models/DomainModels/Plan.swift:16]: Oh I was actually following the data model where the endDate is present:

```
protocol FirebaseAdaptedPlan: FirebaseAdaptedData {
    var tripId: String { get }
    var name: String { get }
    var planType: FirebasePlanType { get }
    var startDate: Date { get }
    var endDate: Date { get }
```

Comment (keanecjy) [reply TourMate/TourMate/Models/DomainModels/Plan.swift:16]: Why does it not have an endDate? I thought that we will be handling startDate and endDate checks that they fit into a trip down the line

**Pull request merged**